### PR TITLE
Prebuilt lut

### PR DESCRIPTION
**Pull request by pgbrodrick**
Overtakes #921 with a fresh write to try and handle the grid appropriately (921 got tangled between config parameters).  Highlevel strategy is to:

- Check prebuilt lut dimensions (inspect_prebuilt_lut_dimensions) -> @jammont, should this live in luts instead of template_construction.py?
- If both heuristic and prebuilt LUT have component, but heuristic range is smaller than LUT, subset LUT
- If a variable is in the heuristic and not in the prebuilt LUT throw an error (TBC, providing the LUT config will subset the heuristic before this check)
- If a variable is in the prebuilt LUT and not in the heuristic, interpolate to the mean.

Still needs testing, and verification around boundary conditions.

Boundary condition points update:

Say: 
  - Heuristic range is [0.60, 1.70]
  - LUT has grid points [0.3, 0.5, 0.8, 1.0, 1.3, 1.5, 1.8, 2.0, 2.3]
  - We should subset the LUT to [0.5, 0.8, 1.0, 1.3, 1.5, 1.8]
 
Alternatively we could interpolate to the upper and lower point, but that would cause a computational hit on load.

### CODE REVIEW

File: isofit/utils/template_construction.py [hunk 1]
```diff
@@ -1543,14 +1602,92 @@ def make_atmosphere_config(
...
+                    interp_value = (lut_min + lut_max) / 2.0
+
+                logging.info(
+                    f"Variable '{lut_dim_name}' is present in prebuilt LUT but not "
+                    f"in heuristic. Will interpolate to {interp_value}"
                 )
-                to_remove.append(gn)
-            else:
-                lut_grid[gn] = get_lut_subset(gc)
+                lut_grid[lut_dim_name] = {"interp": interp_value}
```
Comment (jammont) [inline isofit/utils/template_construction.py]: `lut_grid` isn't the subselection dict, `lut_names` is

File: isofit/utils/template_construction.py [hunk 2]
```diff
@@ -864,12 +864,69 @@ def build_config(
...
+        range_min: minimum value of the aerosol parameter range
+        range_max: maximum value of the aerosol parameter range
+
+    Returns:
+        float: the initial/interpolation value (min + 10% of range)
+    """
+    return (range_max - range_min) / 10.0 + range_min
+
+
+def inspect_prebuilt_lut_dimensions(prebuilt_lut_path: str, ncds: nc.Dataset = None):
```
Comment (jammont) [inline isofit/utils/template_construction.py]: I do like this living in `luts/reader.py`, seems appropriate to have the utility there instead

Comment (jammont) [reply isofit/utils/template_construction.py]: Although this is written for NetCDF. We'll need a generalized function for Zarr

**Review by jammont — APPROVED** [review]
Looks good, just have to make sure the lut subsetting gets assigned to the config's `lut_names`

File: isofit/utils/template_construction.py [hunk 3]
```diff
@@ -1543,20 +1602,120 @@ def make_atmosphere_config(
...
+                    f"in heuristic. Will interpolate to {interp_value}"
                 )
-                to_remove.append(gn)
-            else:
-                lut_grid[gn] = get_lut_subset(gc)
+                lut_grid[lut_dim_name] = {"interp": interp_value}
+
+        # Check for range mismatches - use LUT grid points that encompass heuristic range
+        for dim_name, dim_val in lut_grid.items():
+            if dim_name in lut_dimensions and dim_val is not None:
```
Comment (jammont) [inline isofit/utils/template_construction.py]: This may be simplified using [luts/reader.py#sub](https://github.com/isofit/isofit/blob/dev/isofit/luts/reader.py#L106):

```
    lte: float, default=None
        Select along this dim coordinates that are valued less than or equal to this
    gte: float, default=None
        Select along this dim coordinates that are valued greater than or equal to this
    encompass: bool, default=True
        Change the values of gte/lte such that these values are encompassed using the
        previous/next valid grid point
```

`{"gte": lower, "lte": upper, "encompass": True}` will reassign gte/lte to the grid points the encompass it. `encompass=True` is the parameter default so it doesn't necessarily need to be defined, but it's more explicit if you do. 

Something along the lines of:
```
lut_names = {}
prebuilt = inspect_prebuilt_lut_dimensions(...)
for dim, vals in prebuilt.items():
  if dim not in lut_grid:
    # Squeeze unused dimensions out
    lut_names[dim] = {"interp": (vmin+vmax)/2}
  else:
    # Subselect existing dimensions
    lut_names[dim] = {"gte": vmin, "lte": vmax, "encompass": True}
```
Reader will automatically figure out the encompassing grid points from there

Comment (pgbrodrick) [reply isofit/utils/template_construction.py]: @jammont, check the out, I think it's reduced more along the lines of the suggested.

**Comment by pgbrodrick** [comment]
Now appears to be mostly working.  Presolve does:

```
INFO:2026-06-28,15:40:55 || apply_oe.py:apply_oe() | Path (km): 420.6568072441127
INFO:2026-06-28,15:40:55 || apply_oe.py:apply_oe() | To-sensor azimuth (deg): 142.01654520506906
INFO:2026-06-28,15:40:55 || apply_oe.py:apply_oe() | To-sensor zenith (deg): 5.881249396218373
INFO:2026-06-28,15:40:55 || apply_oe.py:apply_oe() | To-sun azimuth (deg): 134.05646992971134
INFO:2026-06-28,15:40:55 || apply_oe.py:apply_oe() | To-sun zenith (deg): 46.086549374813224
INFO:2026-06-28,15:40:55 || apply_oe.py:apply_oe() | Relative to-sun azimuth (deg): 29.608165165016473
INFO:2026-06-28,15:40:55 || apply_oe.py:apply_oe() | Altitude (km): 418.50706308475776

...

Final lut_names content: {
'AOT550': {'interp': np.float64(0.1305)}, 
'CO2': {'interp': 420.0}, 
'H2OSTR': {'gte': 0.2, 'lte': 5.22, 'encompass': True}, 
'observer_zenith': {'interp': 5.881249396218379}, 
'relative_azimuth': {'interp': 29.608165165016473}, 
'solar_zenith': {'interp': 46.086549374813224}, 
'surface_elevation_km': {'interp': 0.06442344936352495}}
```

Main Solution mostly does, but the solar_zenith falls back to a grid point instead of the interpolated interior, needs examination as to why.  H2OSTR grid needs to be validated to make sure this is encompassing:

```
INFO:2026-06-28,15:52:03 || apply_oe.py:apply_oe() | Full (non-aerosol) LUTs:
INFO:2026-06-28,15:52:03 || apply_oe.py:apply_oe() | Elevation: None
INFO:2026-06-28,15:52:03 || apply_oe.py:apply_oe() | To-sensor zenith: [4.8872 7.9929]
INFO:2026-06-28,15:52:03 || apply_oe.py:apply_oe() | To-sun zenith: None
INFO:2026-06-28,15:52:03 || apply_oe.py:apply_oe() | Relative to-sun azimuth: [ 0.     56.6772]
INFO:2026-06-28,15:52:03 || apply_oe.py:apply_oe() | H2O Vapor: [3.389  3.6376 3.8863 4.135 ]
INFO:2026-06-28,15:52:03 || apply_oe.py:apply_oe() | Writing main configuration file.


Final lut_names content: {
'AOT550': {'gte': np.float64(0.045), 'lte': np.float64(0.9), 'encompass': True}, 
'CO2': {'gte': 380, 'lte': 440, 'encompass': True}, 
'H2OSTR': {'gte': 3.3889999389648438, 'lte': 4.135000228881836, 'encompass': True}, 
'observer_zenith': {'gte': 4.8872, 'lte': 7.9929, 'encompass': True}, 
'relative_azimuth': {'gte': np.int64(2), 'lte': 56.6772, 'encompass': True}, 
'solar_zenith': {'interp': np.float64(37.5)}, 
'surface_elevation_km': {'interp': np.float64(4.0)}}
```

**Comment by pgbrodrick** [comment]
Now getting the desired outcome:

Full solve:

```
INFO:2026-06-28,20:10:10 || apply_oe.py:apply_oe() | Full (non-aerosol) LUTs:
INFO:2026-06-28,20:10:10 || apply_oe.py:apply_oe() | Elevation: None 
INFO:2026-06-28,20:10:10 || apply_oe.py:apply_oe() | To-sensor zenith: [4.8872 7.9929]
INFO:2026-06-28,20:10:10 || apply_oe.py:apply_oe() | To-sun zenith: None 
INFO:2026-06-28,20:10:10 || apply_oe.py:apply_oe() | Relative to-sun azimuth: [ 0.     56.6772]
INFO:2026-06-28,20:10:10 || apply_oe.py:apply_oe() | H2O Vapor: [3.389  3.6376 3.8863 4.135 ]
INFO:2026-06-28,20:10:10 || apply_oe.py:apply_oe() | Writing main configuration file.
INFO:2026-06-28,20:10:10 || template_construction.py:load_climatology() | Loading Climatology
INFO:2026-06-28,20:10:10 || template_construction.py:load_climatology() | Climatology Loaded.  Aerosol State Vector:
{'AOT550': {'bounds': [0.04209, 1.0], 'scale': 1, 'init': 0.137881, 'prior_sigma': 10.0, 'prior_mean': 0.137881}}
Aerosol LUT Grid:
{'AOT550': [0.04209, 0.137881, 0.233672, 0.329463, 0.425254, 0.521045, 0.6168359999999999, 0.712627, 0.808418, 0.9042089999999999, 1.0]}
Aerosol model path:/store/brodrick/repos/isofit-data/aerosol_model.txt
WARNING:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() | Variable 'AOT550' heuristic range [0.0421, 1.0000] constrained to LUT range [0.0450, 0.9000]. Reader will use LUT points that encompass [0.0450, 0.9000]
INFO:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() |   AOT550: Setting lut_grid to 6 LUT points: [0.0450, ..., 0.9000]
INFO:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() |   CO2: Setting lut_grid to 3 LUT points: [380.0000, ..., 440.0000]
INFO:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() |   H2OSTR: Setting lut_grid to 6 LUT points: [3.2000, ..., 4.2000]
INFO:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() |   observer_zenith: Setting lut_grid to 2 LUT points: [0.0000, ..., 10.0000]
WARNING:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() | Variable 'relative_azimuth' heuristic range [0.0000, 56.6772] constrained to LUT range [2.0000, 178.0000]. Reader will use LUT points that encompass [2.0000, 56.6772]
INFO:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() |   relative_azimuth: Setting lut_grid to 3 LUT points: [2.0000, ..., 90.0000]
INFO:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() | Variable 'solar_zenith' is present in prebuilt LUT but not in heuristic. Will interpolate to 46.0865 (scene mean from template)
INFO:2026-06-28,20:10:10 || template_construction.py:make_atmosphere_config() | Variable 'surface_elevation_km' is present in prebuilt LUT but not in heuristic. Will interpolate to 0.0644 (scene mean from template)
```

**Comment by evan-greenbrg** [comment]
The test:

With the same EMIT granule, generate configs via:
1. `dev`
2. `This PR` (no `--prebuilt_lut`)
3. `This PR` (with `--prebuilt_lut` pointing to `lut.nc`)

Each case gives me equivalent LUT grids:
AOD: `[0.04209 - 1.0]`
H2O: `[0.2- 5.346]`
observer_zenith: `[8.2748-10.56]`
relative_azimuth: `[0.0-59.3612]`
surface_elevation_km: `[0.0022-0.6529]`

The only difference I see in the lut_grids is the rounding in the H2O grid. The --prebuilt rounds to 2 decimals, while the other run paths round to 4.

**Comment by pgbrodrick** [comment]
Ready to merge with the rest of PRs.
